emulab-devel-new/emulab-devel/clientside/os/syncd/emulab-sync.1
.\"
.\" Copyright (c) 2004 University of Utah and the Flux Group.
.\" 
.\" {{{EMULAB-LICENSE
.\" 
.\" This file is part of the Emulab network testbed software.
.\" 
.\" This file is free software: you can redistribute it and/or modify it
.\" under the terms of the GNU Affero General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or (at
.\" your option) any later version.
.\" 
.\" This file is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public
.\" License for more details.
.\" 
.\" You should have received a copy of the GNU Affero General Public License
.\" along with this file.  If not, see <http://www.gnu.org/licenses/>.
.\" 
.\" }}}
.\"
.TH EMULAB-SYNC 1 "April 5, 2004" "Emulab" "Emulab Commands Manual"
.OS
.SH NAME
emulab-sync \- Simple distributed synchronization client.
.SH SYNOPSIS
.BI emulab-sync
[\fB-hV\fR]
[\fB-n \fIname\fR]
[\fB-d\fR]
[\fB-s \fIserver\fR]
[\fB-p \fIport-number\fR]
[\fB-i \fIcount\fR]
[\fB-a\fR]
[\fB-u\fR]
[\fB-e \fIerror-number\fR]
[\fB-m\fR]
.SH DESCRIPTION
The
.B emulab-sync
utility is used to wait for other programs in your experiment to reach some
specific point before proceeding forward.  This utility and its server,
.B emulab-syncd\fR,
implement this synchronization through a simple counter based barrier.  The
counter is increased by
.B emulab-sync
when using the
.B -i
option and then decremented by one on each regular use.  Once the counter
reaches zero, all of the waiters will return and execution can proceed.
.P
Multiple synchronizations can take place at the same time by using the
.B -n
option to specify a separate name for each barrier.  For example, scripts on
nodeA and nodeB can be waiting on a barrier named "foo" while (other) scripts
on nodeA and nodeC can be waiting on a barrier named "bar." You may reuse an
existing barrier (including the default barrier) once it has been released (all
clients arrived and woken up).
.P
Basic error reporting is also supported using the
.B -e
option and the program's exit code.  For example, if a client encountered an
error and would be unable to continue after crossing the barrier, it can signal
its peers by using the
.B -e
option with a non-zero error code.  Then, when all of the clients have arrived,
they will wake up and exit with this non-zero error code.  If multiple peers
encounter errors, the maximum of all of the errors will be returned.
.P
Available options:
.P
.TP
\fB-h
Print out a usage message.
.TP
\fB-V
Print out version information and exit.
.TP
\fB-n \fIname
The name of the barrier, must be less than 64 bytes long.  (Default: barrier)
.TP
\fB-d
Turn on debugging.
.TP
\fB-s \fIserver
Specify the host where the
.B emulab-syncd
server is running.  Normally, you do not need to use this option since Emulab
will automatically use the experiment's default sync server.  The experiment's
sync server is picked at random or specified by the
.B tb-set-sync-server
NS command when creating the experiment.
.TP
\fB-p \fIport-number
Specify the port number the server is listening on.  (Default: 16534)
.TP
\fB-i \fIcount
Increase the number of waiters by
.I count
waiters.
.B NOTE\fR:
The number of waiters should not include this call, unless you are using the
.B -a
option and intend to do another sync later on.
.TP
\fB-a
Return immediately after increasing the barrier value.  Can only be used 
with the
.B -i
option.
.TP
\fB-u
Use UDP to contact the server.  (Default: TCP)
.TP
\fB-e \fIerror-number

Specify the error code for this client.  The value must be greater than zero
(indicating no error) and less than 240 because we reserve the range [240, 256)
for internal use.  The error codes for all of the clients are collected by the
server and the maximum is returned by the clients when they exit.  (Default: 0)
.TP
\fB-m
Check if this node is where the default sync server is running and then
immediately exit.  If this node
.I is
running the server, the return value is zero, otherwise it is one.  This mode
is useful for detecting the "master" node when the experimental nodes all run
the same startup script.
.SH RETURN VALUES
.TP
A \fB-e \fRerror value
If the barrier was crossed, but atleast one of the clients reported a non-zero
error code.
.TP
240
If the server received a SIGHUP from a user, causing all of the barriers 
to be cleared and the clients released.
.TP
76
If there was an error while communicating with the server.
.TP
68
If the server name lookup failed.
.TP
64
If there was an invalid command line argument.
.TP
1
If the
.B -m
option was used and this node is not the master.
.TP
0
If the barrier was successfully crossed without error or the
.B -m
option was used and this node is the master.
.SH EXAMPLES
.PP
To synchronize one machine with another:
.PP
.RS
[gob@bluth1 ~] emulab-sync -i 1
.P
[michael@bluth2 ~] emulab-sync
.RE
.PP
To perform two different synchronizations at the same time using barriers named
"staircar" and "prison":
.PP
.RS
[marta@bluth1 ~] emulab-sync -i 1 -n staircar
.P
[michael@bluth2 ~] emulab-sync -n staircar
.P
[gob@bluth1 ~] emulab-sync -i 1 -n prison
.P
[george@bluth2 ~] emulab-sync -n prison
.RE
.PP
To synchronize and report an error code of "100":
.PP
.RS
.PD 0
[buster@bluth1 ~] emulab-sync -i 1 -e 100; echo $?
.P
100
.PD
.P
.PD 0
[annyong@bluth2 ~] emulab-sync -e 0; echo $?
.P
100
.PD
.RE
.PP
To use a single script to synchronize N nodes:
.PP
.RS
.PD 0
#! /bin/sh

# Get number of nodes to synchronize against.
.P
node_count=${1}

# Synchronize first.
.P
\fBif\fR emulab-sync -m; \fBthen\fR
.P
    # This is the master node, initialize the
.P
    # barrier to N - 1 since the master node
.P
    # is not counted.
.P
    emulab-sync -i `expr ${node_count} - 1`
.P
\fBelse\fR
.P
    # This is a regular node, just do a regular
.P
    # sync.
.P
    emulab-sync
.P
\fBfi\fR

# ... then do something interesting.
.RE
.PD
.SH FILES
.TP
/var/emulab/boot/syncserver
Contains the name and port number of the experiment's default sync server.
.SH NOTES
If a client using TCP is unable to connect to the server, it will continue to
retry at five second intervals.  Clients using UDP do not support such
reliability, so be careful.
.P
If you terminate a TCP client before it returns, the server will recognize the
closed socket and adjust the counter so that it appears that the call 
never occurred.
.SH BUGS
Despite the above efforts, it is still possible for a barrier's counter to end
up with an unknown value.  Typical symptoms of this condition are clients that
refuse to wakeup despite all of them checking in.  At this point it is probably
best to use a different barrier name or you can clear all of the barriers by
sending a SIGHUP to the experiment's
.B emulab-syncd\fR.
.SH SEE ALSO
emulab-syncd(1)
.SH AUTHOR
The Emulab project at the University of Utah.
.SH NOTES
The Emulab project can be found on the web at
.IR http://www.emulab.net


emulab-devel-new/emulab-devel/clientside/os/syncd/emulab-sync.c
/*
 * Copyright (c) 2000-2012 University of Utah and the Flux Group.
 * 
 * {{{EMULAB-LICENSE
 * 
 * This file is part of the Emulab network testbed software.
 * 
 * This file is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at
 * your option) any later version.
 * 
 * This file is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this file.  If not, see <http://www.gnu.org/licenses/>.
 * 
 * }}}
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <sys/un.h>
#include <sys/fcntl.h>
#include <stdio.h>
#include <errno.h>
#include <syslog.h>
#include <unistd.h>
#include <signal.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <time.h>
#include <assert.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <setjmp.h>
#include "decls.h"

#ifndef NO_EMULAB
#include "config.h"
#include "tmcc.h"
#endif

#ifndef SYNCSERVER
#define SYNCSERVER	"/var/emulab/boot/syncserver"
#endif

static int	debug = 0;

/* Forward decls */
static int	doudp(barrier_req_t *, struct in_addr, int);
static int	dotcp(barrier_req_t *, struct in_addr, int);

/* For connect timeouts. */
static jmp_buf	deadline;
static int	deadsyncdflag;
static void	deadsyncd(int sig);

char *usagestr = 
 "usage: emulab-sync [options]\n"
 " -h              Display this message\n"
 " -V              Print version information and exit\n"
 " -n name         Optional barrier name; must be less than 64 bytes long\n"
 " -d              Turn on debugging\n"
 " -s server       Specify a sync server to connect to\n"
 " -p portnum      Specify a port number to connect to\n"
 " -i count        Increase the barrier value by 'count' waiters\n"
 " -a              Master returns immediately when initializing barrier\n"
 " -u              Use UDP instead of TCP\n"
 " -e errornum     Signal an error (must be zero or >= 32)\n"
 " -m              Test whether or not this is the 'master' node\n"
 "\n";

void
usage()
{
	fprintf(stderr, "%s", usagestr);
	exit(64);
}

int
main(int argc, char **argv)
{
	int			n, ch, errval = 0;
	struct hostent		*he;
	unsigned char		**hal;
	FILE			*fp;
	struct in_addr		serverip;
	int			portnum = 0;
	int			useudp  = 0;
	char			*server = NULL;
	int			initme  = 0;
	int			nowait  = 0;
	char			*barrier= DEFAULT_BARRIER;
	barrier_req_t		barrier_req;
#ifndef NO_EMULAB
	extern char		build_info[];
#endif
	int			mastercheck = 0;
	
	while ((ch = getopt(argc, argv, "hVds:p:ui:n:ae:m")) != -1)
		switch(ch) {
		case 'a':
			nowait++;
			break;
		case 'd':
			debug++;
			break;
		case 'p':
			if (sscanf(optarg, "%d", &portnum) == 0) {
				fprintf(stderr,
					"Error: -p value is not a number: "
					"%s\n",
					optarg);
				usage();
			}
			else if ((portnum <= 0) || (portnum >= 65536)) {
				fprintf(stderr,
					"Error: -p value is not between "
					"0 and 65536: %d\n",
					portnum);
				usage();
			}
			break;
		case 'i':
			if (sscanf(optarg, "%d", &initme) == 0) {
				fprintf(stderr,
					"Error: -i value is not a number: "
					"%s\n",
					optarg);
				usage();
			}
			else if (initme <= 0) {
				fprintf(stderr,
					"Error: -i value must be greater "
					"than zero: %s\n",
					optarg);
				usage();
			}
			break;
		case 'n':
			if (strlen(optarg) == 0) {
				fprintf(stderr, "Error: -n value is empty\n");
				usage();
			}
			else if (strlen(optarg) >= 64) {
				fprintf(stderr,
					"Error: -n value is too long: %s\n",
					optarg);
				usage();
			}
			else {
				barrier = optarg;
			}
			break;
		case 's':
			if (strlen(optarg) == 0) {
				fprintf(stderr, "Error: -s value is empty\n");
				usage();
			}
			else {
				server = optarg;
			}
			break;
		case 'u':
			useudp = 1;
			break;
		case 'e':
			if (sscanf(optarg, "%d", &errval) == 0) {
				fprintf(stderr,
					"Error: -e value is not a number: "
					"%s\n",
					optarg);
				usage();
			}
			else if ((errval < 0) ||
				 (errval >= SERVER_ERROR_BASE)) {
				fprintf(stderr,
					"Error: -e value must be zero or "
					"less than %d\n",
					SERVER_ERROR_BASE);
				usage();
			}
			break;
		case 'V':
#ifndef NO_EMULAB
			fprintf(stderr, "%s\n", build_info);
#else
                        fprintf(stderr, "Built in standalone mode\n");
#endif
			exit(0);
			break;
		case 'h':
			fprintf(stderr, "%s", usagestr);
			exit(0);
			break;
		case 'm':
			mastercheck = 1;
			break;
		default:
			usage();
		}

	if (nowait && (initme == 0)) {
		fprintf(stderr,	"Error: -a requires the -i option\n");
		usage();
	}
	
	argv += optind;
	argc -= optind;

	if (argc) {
		fprintf(stderr,
			"Error: Unrecognized command line arguments: %s ...\n",
			argv[0]);
		usage();
	}

	/*
	 * Look for the syncserver access file.
	 */
	if (!server &&
	    (access(SYNCSERVER, R_OK) == 0) &&
	    ((fp = fopen(SYNCSERVER, "r")) != NULL)) {
		char	buf[BUFSIZ], *bp;
		
		if (fgets(buf, sizeof(buf), fp)) {
			if ((bp = strchr(buf, '\n')))
				*bp = 0;

			/*
			 * Look for port spec
			 */
			if ((bp = strchr(buf, ':'))) {
				*bp++   = 0;
				portnum = atoi(bp);
			}
			server = strdup(buf);
		}
		fclose(fp);
	}
	if (!server) {
		fprintf(stderr,
			"Error: Could not deduce the name of the syncserver!\n"
			"You may be running in an older experiment, you\n"
			"will have to manually start the emulab-syncd and\n"
			"use the -s option, or, you can recreate your\n"
			"experiment.\n\n");
		usage();
	}

	/*
	 * Map server to IP.
	 */
	he = gethostbyname(server);
	if (he)
		memcpy((char *)&serverip, he->h_addr, he->h_length);
	else {
		fprintf(stderr, "gethostbyname(%s) failed\n", server); 
		exit(68);
	}
	if (debug) {
		printf( "server name %s, addr(s)", server);
		for (hal = (unsigned char **)he->h_addr_list; *hal; hal++)
			printf(" %u.%u.%u.%u\n", 
			       (*hal)[0], (*hal)[1],(*hal)[2], (*hal)[3]);
	}

	if (mastercheck) {
		/* Compare server IP with the current host IP addresses.
		 * (Windows nodes all have the physical node name.)
		 */
		char nodename[BUFSIZ];
		gethostname(nodename, sizeof(nodename));
		he = gethostbyname(nodename); /* Internal retval storage. */
		if (!he) {
			fprintf(stderr, "gethostbyname(%s) failed\n", nodename); 
			exit(68);
		}
		if (debug) {
			printf( "node name %s, addr(s)", nodename);
			for (hal = (unsigned char **)he->h_addr_list; *hal; hal++)
				printf(" %u.%u.%u.%u\n",
				       (*hal)[0], (*hal)[1],(*hal)[2], (*hal)[3]);
		}
		/* Host is multi-homed.  Compare to ALL of its addresses. */
		for (hal = (unsigned char **)he->h_addr_list; *hal; hal++)
			if (memcmp((char *)&serverip, *hal, he->h_length) == 0)
				exit(0);
		exit(1);
	}

	/*
	 * Lookup the port.
	 */
	if (portnum == 0) {
#ifndef NO_EMULAB
		char nodename[BUFSIZ];
		int  rc;

		while (1) {
			rc = PortLookup(SERVER_SERVNAME, nodename,
					sizeof(nodename), &portnum);

			/* Got the port; we do not care about the nodename */
			if (rc == 0 && portnum != 0) {
				if (debug) {
					printf("Sync Server at %s:%d\n",
					       nodename, portnum);
				}
				break;
			}

			if (rc == 0 && portnum == 0) {
				fprintf(stderr,
					"Could not lookup service with tmcd!");
				exit(1);
			}

			/* Delay */
			sleep(5);
		}
#else
                fprintf(stderr, "No port specified\n");
                exit(1);
#endif
	}
	
	/*
	 * Build up the request structure to send to the server.
	 */
	bzero(&barrier_req, sizeof(barrier_req));
	strcpy(barrier_req.name, barrier);
	if (initme) {
		barrier_req.request = BARRIER_INIT;
		barrier_req.count   = initme;
		if (nowait)
			barrier_req.flags = BARRIER_INIT_NOWAIT;
	}
	else
		barrier_req.request = BARRIER_WAIT;
	barrier_req.error = errval;
	
	if (useudp)
		n = doudp(&barrier_req, serverip, portnum);
	else
		n = dotcp(&barrier_req, serverip, portnum);
	
	if (n < 0)
		exit(76); /* Protocol failure */
	else
		exit(n); /* User defined error. */

	assert(0); /* Not reached. */
}

/*
 * TCP version, which uses ssl if compiled in.
 */
static int
dotcp(barrier_req_t *barrier_reqp, struct in_addr serverip, int portnum)
{
	int			sock, n, cc;
	struct sockaddr_in	name;
	char			*bp, buf[BUFSIZ];
	
	while (1) {
		/* Create socket from which to read. */
		sock = socket(AF_INET, SOCK_STREAM, 0);
		if (sock < 0) {
			perror("creating stream socket:");
			return -1;
		}

		/* Create name. */
		name.sin_family = AF_INET;
		name.sin_addr   = serverip;
		name.sin_port   = htons(portnum);

		/* For alarm. */
		deadsyncdflag = 0;
		signal(SIGALRM, deadsyncd);
		if (setjmp(deadline)) {
			alarm(0);
			signal(SIGALRM, SIG_DFL);
			goto refused;
		}
		alarm(5);

		cc = connect(sock, (struct sockaddr *) &name, sizeof(name));
		/*
		 * If we get the alarm now, no problem. We will close the
		 * connection and try again after a short wait. The server
		 * will recongnize this for what is, not recording any state
		 * until we actually send it the request below (write to sock).
		 */
		alarm(0);
		signal(SIGALRM, SIG_DFL);

		/* Success */
		if (cc == 0)
			break;

		if (errno != ECONNREFUSED && errno != ETIMEDOUT) {
			perror("connecting stream socket");
			close(sock);
			return -1;
		}
	refused:
		close(sock);
		if (debug) 
			fprintf(stderr, "Connection refused. Waiting ...\n");
		sleep(5);
	}

	n = 1;
	if (setsockopt(sock, SOL_SOCKET, SO_KEEPALIVE, &n, sizeof(n)) < 0) {
		perror("setsockopt SO_KEEPALIVE");
		goto bad;
	}

	/*
	 * Write the command to the socket and wait for the response.
	 */
	bp = (char *) barrier_reqp;
	n  = sizeof(*barrier_reqp);
	while (n) {
		if ((cc = write(sock, bp, n)) <= 0) {
			if (cc < 0) {
				perror("Writing to socket:");
				goto bad;
			}
			fprintf(stderr, "write aborted");
			goto bad;
		}
		bp += cc;
		n  -= cc;
	}

	bp = buf;
	n = 0;
	memset(buf, 0, sizeof(buf));
	while (1) {
		if ((cc = read(sock, bp, sizeof(buf) - n)) <= 0) {
			if (cc < 0) {
				perror("Reading from socket:");
				goto bad;
			}
			break;
		}
		else {
			bp += cc;
			n += cc;
		}
	}
	close(sock);

	if (n == sizeof(int32_t)) {
		return *((int32_t *)buf);
	}
	else {
		return -1;
	}
 bad:
	close(sock);
	return -1;
}

/*
 * Not very robust ...
 */
static int
doudp(barrier_req_t *barrier_reqp, struct in_addr serverip, int portnum)
{
	int			sock, n, cc;
	socklen_t		length;
	struct sockaddr_in	name, client;
	char			buf[BUFSIZ];

	/* Create socket from which to read. */
	sock = socket(AF_INET, SOCK_DGRAM, 0);
	if (sock < 0) {
		perror("creating dgram socket:");
		return -1;
	}

	/* Create name. */
	name.sin_family = AF_INET;
	name.sin_addr   = serverip;
	name.sin_port   = htons(portnum);

	/*
	 * Write the command to the socket and wait for the response
	 */
	n  = sizeof(*barrier_reqp);
	cc = sendto(sock, barrier_reqp,
		    n, 0, (struct sockaddr *)&name, sizeof(name));
	if (cc != n) {
		if (cc < 0) {
			perror("Writing to socket:");
			return -1;
		}
		fprintf(stderr, "short write (%d != %d)\n", cc, n);
		return -1;
	}

	cc = recvfrom(sock, buf, sizeof(buf) - 1, 0,
		      (struct sockaddr *)&client, &length);

	if (cc < 0) {
		perror("Reading from socket:");
		return -1;
	}
	close(sock);

	if (cc == sizeof(int32_t)) {
		return *((int32_t *)buf);
	}
	else {
		return -1;
	}
}

void
deadsyncd(int sig)
{
	deadsyncdflag = 1;
	longjmp(deadline, 1);
}
